### PR TITLE
Bloom Filter

### PR DESCRIPTION
**Pull request by subsymbolic**
Task/Issue URL: https://app.asana.com/0/72649045549333/744008606284177
Tech Design URL: https://app.asana.com/0/72649045549333/727813825116580

Description:
Bloom filter with djb2 and sdbm hashing. It is a loose C++ port of the js library at https://github.com/cry/jsbloom.

Testing:
Note you will need cmake installed on your computer, if you use homebrew you can install it with `brew install cmake`
- Run the `run_tests.sh` script to run the unit tests

### CODE REVIEW

File: src/BloomFilter.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/*
+ Bloom filter with djb2 and sdbm hashing. It is a loose C++ port of
+ the js library at https://github.com/cry/jsbloom
+ */
+class BloomFilter {
+
+private:
+    unsigned int hashRounds;
+    std::vector<bool> bloomVector;
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply src/BloomFilter.hpp]: Bool vectors are memory optimized with multiple bits packed into a single byte so unlike other vectors their memory usage is not the element size * number of elements.

[the remark this answers is not in the document]
Comment (brindy) [reply src/BloomFilter.hpp]: Oh cool, that's super handy.

File: src/BloomFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+#include <iterator>
+#include <stdio.h>
+#include <cstdio>
+#include <fstream>
+#include <math.h>
+#include "BloomFilter.hpp"
+
+using namespace std;
+
+// Forward declarations of free functions
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply src/BloomFilter.cpp]: I'll go with static. I hate exposing superfluous stuff in the .h file.

File: src/BloomFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    hashRounds = calculateHashRounds((unsigned int) bloomVector.size(), maxItems);
+}
+
+BloomFilter::BloomFilter(BinaryInputStream& in, unsigned int maxItems)
+{
+    bloomVector = readVectorFromStream(in);
+    hashRounds = calculateHashRounds((unsigned int) bloomVector.size(), maxItems);
+}
+
+unsigned int calculateHashRounds(unsigned int size, unsigned int maxItems)
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply src/BloomFilter.cpp]: I'm felxible on this althoug I like to define them as soon as I need them. I see a lot of people define them just before they are used (which avoids forward declarations) but I find it nicer to define them after when there's more context, in a similar way to how we define private methods in our other projects. More a style thing so not a biggie either way - let's discus is and some to some consensus on the style.

File: src/BloomFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+void unpackIntoVector(vector<bool>& bloomVector,
+                      const size_t offset,
+                      const size_t bitsInThisBlock,
+                      BinaryInputStream& in);
+
+
+// Implementation
+
+BloomFilter::BloomFilter(unsigned int maxItems, double targetProbability)
+{
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply src/BloomFilter.cpp]: I changed my mind on this and ended up using the came style we already use in other projects/lanbguages.

**Comment by subsymbolic** [comment]
@brindy this is ready for re-review. Did you want to catch up over zoom to iron out any final style discussions?

**Pull request merged**